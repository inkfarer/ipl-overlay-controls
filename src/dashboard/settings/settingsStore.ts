import { NodeCGBrowser } from 'nodecg/browser';
import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        }
    }
});
