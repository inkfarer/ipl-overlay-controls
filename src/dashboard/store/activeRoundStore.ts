import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
