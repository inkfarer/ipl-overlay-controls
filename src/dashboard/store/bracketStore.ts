import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
