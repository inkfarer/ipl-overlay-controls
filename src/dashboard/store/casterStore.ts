import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
