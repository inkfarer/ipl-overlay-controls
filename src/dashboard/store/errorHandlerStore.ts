import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
