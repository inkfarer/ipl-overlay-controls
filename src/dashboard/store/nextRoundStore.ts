import { NextRound } from 'schemas';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';
import { defineStore } from 'pinia';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const useNextRoundStore = defineStore('nextRound', {
    state: () => ({
        nextRound: null
    } as NextRoundStore),
    actions: {
        setShowOnStream(newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        },
        beginNextMatch(request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(request: SetNextRoundRequest): void {
            nodecg.sendMessage('setNextRound', request);
        }
    }
});
