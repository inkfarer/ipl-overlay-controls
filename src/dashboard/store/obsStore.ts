import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
