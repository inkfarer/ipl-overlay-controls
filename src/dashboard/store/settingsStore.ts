import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';
import { Locale } from 'types/enums/Locale';
import { perGameData } from '../../helpers/gameData/gameData';
import { InterfaceLocale } from 'types/enums/InterfaceLocale';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    getters: {
        translatedModeName: state =>
            (mode: string) => (perGameData[state.runtimeConfig.gameVersion]
                .modes[state.runtimeConfig.locale] as Record<string, string>)[mode],
        translatedStageName: state =>
            (stage: string) => (perGameData[state.runtimeConfig.gameVersion]
                .stages[state.runtimeConfig.locale] as Record<string, string>)[stage],
    },
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        },
        setLocale(newValue: Locale): Promise<void> {
            return nodecg.sendMessage('setLocale', newValue);
        },
        setInterfaceLocale(newValue: `${InterfaceLocale}`) {
            runtimeConfig.value.interfaceLocale = newValue;
        }
    }
});
