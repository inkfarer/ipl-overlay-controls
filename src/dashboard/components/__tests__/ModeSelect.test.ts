import { config, shallowMount } from '@vue/test-utils';
import ModeSelect from '../ModeSelect.vue';
import { useSettingsStore } from '../../store/settingsStore';
import { GameVersion } from 'types/enums/gameVersion';
import { Locale } from 'types/enums/Locale';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import { IplSelect } from '@iplsplatoon/vue-components';
import { InterfaceLocale } from 'types/enums/InterfaceLocale';

describe('ModeSelect', () => {
    let pinia: TestingPinia;

    beforeEach(() => {
        pinia = createTestingPinia();

        config.global.plugins = [pinia];

        useSettingsStore().runtimeConfig = {
            gameVersion: GameVersion.SPLATOON_2,
            locale: Locale.EN,
            interfaceLocale: InterfaceLocale.EN,
            activeGraphicsBundles: []
        };
    });

    it('emits expected event when new option is selected', () => {
        const wrapper = shallowMount(ModeSelect);

        wrapper.findComponent<typeof IplSelect>('ipl-select-stub').vm.$emit('update:modelValue', 'New Stage');

        expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['New Stage']);
    });

    it.each(Object.values(Locale))('adds expected options to select when language is %s', locale => {
        useSettingsStore().runtimeConfig.locale = locale;

        const wrapper = shallowMount(ModeSelect);

        expect(wrapper.findComponent<typeof IplSelect>('ipl-select-stub').vm.options).toMatchSnapshot();
    });
});
