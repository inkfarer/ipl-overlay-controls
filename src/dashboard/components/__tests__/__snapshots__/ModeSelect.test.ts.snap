// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModeSelect adds expected options to select when language is DE 1`] = `
[
  {
    "name": "Herrschaft",
    "value": "Splat Zones",
  },
  {
    "name": "Muschelchaos",
    "value": "Clam Blitz",
  },
  {
    "name": "Operation Goldfisch",
    "value": "Rainmaker",
  },
  {
    "name": "Revierkampf",
    "value": "Turf War",
  },
  {
    "name": "Turmkommando",
    "value": "Tower Control",
  },
  {
    "name": "Unbekannte Kampfart",
    "value": "Unknown Mode",
  },
]
`;

exports[`ModeSelect adds expected options to select when language is EN 1`] = `
[
  {
    "name": "Clam Blitz",
    "value": "Clam Blitz",
  },
  {
    "name": "Rainmaker",
    "value": "Rainmaker",
  },
  {
    "name": "Splat Zones",
    "value": "Splat Zones",
  },
  {
    "name": "Tower Control",
    "value": "Tower Control",
  },
  {
    "name": "Turf War",
    "value": "Turf War",
  },
  {
    "name": "Unknown Mode",
    "value": "Unknown Mode",
  },
]
`;

exports[`ModeSelect adds expected options to select when language is EU_FR 1`] = `
[
  {
    "name": "Défense de zone",
    "value": "Splat Zones",
  },
  {
    "name": "Expédition risquée",
    "value": "Tower Control",
  },
  {
    "name": "Guerre de territoire",
    "value": "Turf War",
  },
  {
    "name": "Mission Bazookarpe",
    "value": "Rainmaker",
  },
  {
    "name": "Mode à déterminer",
    "value": "Unknown Mode",
  },
  {
    "name": "Pluie de palourdes",
    "value": "Clam Blitz",
  },
]
`;
